### PR TITLE
Allow using custom noise covariance matrices

### PR DESCRIPTION
**Pull request by hoechenberger**
Closes #531

### Before merging …

- [x] Changelog has been updated (`docs/source/changes.md`)

### CODE REVIEW

File: docs/source/changes.md [hunk 1]
```diff
@@ -137,6 +138,9 @@ authors:
 - The covariance matrix and corresponding SVD figures, as well as whitened
   evoked data plots are now added to the reports.
   ({{ gh(532) }} by {{ authors.hoechenberger }})
+- The [`noise_cov`][config.noise_cov] option can now be set to a function to
+  allow working with arbitrarily-generated covariance matrices.
+  ({{ gh(xxx) }} by {{ authors.hoechenberger }} and {{ authors.agramfort }}
```
Comment (hoechenberger) [inline docs/source/changes.md:143]:
```suggestion
  ({{ gh(535) }} by {{ authors.hoechenberger }} and {{ authors.agramfort }}
```

File: config.py [hunk 2]
```diff
@@ -1567,8 +1577,18 @@ def mri_landmarks_kind(bids_path):
     ```python
     noise_cov = 'ad-hoc'
     ```
-"""
 
+    Use a custom covariance:
+    ```python
+    def gen_custom_cov(bids_path):
+        bids_path.update(task='rest', run=None, suffix='meg')
+        raw_rest = mne_bids.read_raw_bids(bids_path)
```
Comment (agramfort) [inline config.py:1585]:
```suggestion
        bp = bids_path.copy().update(task='rest', run=None, suffix='meg')
        raw_rest = mne_bids.read_raw_bids(bp)
```

File: scripts/source/_03_make_cov.py [hunk 3]
```diff
@@ -80,9 +80,45 @@ def compute_cov_from_empty_room(cfg, subject, session):
...
+        root=cfg.deriv_root,
+        check=False
+    )
+    cov_fname = evoked_bids_path.copy().update(suffix='cov')
+
+    msg = (f'Retrieving noise covariance matrix from custom user-supplied '
+           f'function, Output: {cov_fname}')
+    logger.info(**gen_log_kwargs(message=msg, subject=subject,
+                                 session=session))
+    cov = custom_func(BIDSPath)
```
Comment (agramfort) [inline scripts/source/_03_make_cov.py:110]:
```suggestion
    cov = custom_func(evoked_bids_path)
```

File: scripts/source/_03_make_cov.py [hunk 4]
```diff
@@ -40,7 +40,7 @@ def compute_cov_from_epochs(cfg, subject, session, tmin, tmax):
         processing = 'clean'
 
     epo_fname = bids_path.copy().update(processing=processing, suffix='epo')
-    cov_fname = bids_path.copy().update(suffix='cov')
+    cov_fname = bids_path.copy().update(task='noise', suffix='cov')
```
[the remark this answers is not in the document]
Comment (hoechenberger) [reply scripts/source/_03_make_cov.py:43]: Will revert

File: config.py [hunk 5]
```diff
@@ -1567,8 +1577,18 @@ def mri_landmarks_kind(bids_path):
...
+    Use a custom covariance:
+    ```python
+    def gen_custom_cov(bids_path):
+        bp = bids_path.copy().update(task='rest', run=None, suffix='meg')
+        raw_rest = mne_bids.read_raw_bids(bp)
+        raw.crop(tmin=5, tmax=60)
+        cov = mne.compute_raw_covariance(raw, rank='info')
+        return cov
+
+    noise_cov = gen_custom_cov  # no parentheses here!
```
Comment (agramfort) [inline config.py:1590]:
```suggestion
    def noise_cov(bids_path):
        bp = bids_path.copy().update(task='rest', run=None, suffix='meg')
        raw_rest = mne_bids.read_raw_bids(bp)
        raw.crop(tmin=5, tmax=60)
        cov = mne.compute_raw_covariance(raw, rank='info')
        return cov
```

File: scripts/source/_03_make_cov.py [hunk 6]
```diff
@@ -80,12 +81,49 @@ def compute_cov_from_empty_room(cfg, subject, session):
...
+        processing='custom', suffix='cov'
+    )
+
+    msg = (f'Retrieving noise covariance matrix from custom user-supplied '
+           f'function, Output: {cov_fname.fpath.name}')
+    logger.info(**gen_log_kwargs(message=msg, subject=subject,
+                                 session=session))
+    cov = config.noise_cov(evoked_bids_path)
+    # assert isinstance(cov, mne.Covariance)
+    # cov.save(cov_fname, overwrite=True)
```
Comment (agramfort) [inline scripts/source/_03_make_cov.py:114]: to cleanup

File: scripts/source/_03_make_cov.py [hunk 7]
```diff
@@ -101,21 +142,27 @@ def get_config(
         rec=config.rec,
         space=config.space,
         proc=config.proc,
-        noise_cov=config.noise_cov,
+        # noise_cov=config.noise_cov,
```
Comment (hoechenberger) [inline scripts/source/_03_make_cov.py:145]:
```suggestion
```

**Comment by hoechenberger** [comment]
… half a dozen force-pushes later … Let's hope this comes back green now! All comments addressed.

**Comment by hoechenberger** [comment]
Rendered report:
https://output.circle-artifacts.com/output/job/8823de5d-4f4d-4b1a-aa8a-471fb43e2da3/artifacts/0/reports/ds000248/sub-01_task-audiovisual_report.html

Example whitened ERF from the report:

Screen Shot 2022-04-14 at 20 38 54

Seems to work!

**Pull request merged**